From edf9dbb708ce29bd386bcc3d23cf70e93a852a7e Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Fri, 26 Mar 2021 15:07:27 +0300
Subject: [PATCH backport v5.10 30/32] platform/mellanox: mlxreg-io: Add field
 for secured access

Extended structure 'mlxreg_core_data' with 'secured' field. The purpose
of this field is to set restrict access to some attributes, required
authorization access.

Add validation of new 'secured' field for write access for particular
attributes, in case this filed is set for this attribute.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/platform/mellanox/mlxreg-io.c | 6 ++++++
 include/linux/platform_data/mlxreg.h  | 2 ++
 2 files changed, 8 insertions(+)

diff --git a/drivers/platform/mellanox/mlxreg-io.c b/drivers/platform/mellanox/mlxreg-io.c
index 43e9228a8fd3..7799ae2c3ffd 100644
--- a/drivers/platform/mellanox/mlxreg-io.c
+++ b/drivers/platform/mellanox/mlxreg-io.c
@@ -15,6 +15,7 @@
 #include <linux/platform_data/mlxreg.h>
 #include <linux/platform_device.h>
 #include <linux/regmap.h>
+#include <linux/security.h>
 
 /* Attribute parameters. */
 #define MLXREG_IO_ATT_SIZE	10
@@ -137,6 +138,11 @@ mlxreg_io_attr_store(struct device *dev, struct device_attribute *attr,
 	u32 input_val, regval;
 	int ret;
 
+	if (data->secured) {
+		if (!capable(CAP_SYS_ADMIN) || security_locked_down(LOCKDOWN_DEV_MEM))
+			return -EPERM;
+	}
+
 	if (len > MLXREG_IO_ATT_SIZE)
 		return -EINVAL;
 
diff --git a/include/linux/platform_data/mlxreg.h b/include/linux/platform_data/mlxreg.h
index 4b4149256039..4bb15a60840f 100644
--- a/include/linux/platform_data/mlxreg.h
+++ b/include/linux/platform_data/mlxreg.h
@@ -113,6 +113,7 @@ struct mlxreg_hotplug_device {
  * @attached: true if device has been attached after good health indication;
  * @regnum: number of registers occupied by multi-register attribute;
  * @slot: slot number, at which device is located;
+ * @secured: if set indicates that entry access is secured;
  */
 struct mlxreg_core_data {
 	char label[MLXREG_CORE_LABEL_MAX_SIZE];
@@ -130,6 +131,7 @@ struct mlxreg_core_data {
 	bool attached;
 	u8 regnum;
 	u8 slot;
+	u8 secured;
 };
 
 /**
-- 
2.11.0

